pkg/api/transport_custom.go

```diff
@@ -0,0 +1,223 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"net/http/httputil"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/elastic/cloud-sdk-go/pkg/api/mock"
+	"github.com/elastic/cloud-sdk-go/pkg/models"
+	"github.com/elastic/cloud-sdk-go/pkg/multierror"
+	"github.com/elastic/cloud-sdk-go/pkg/util/ec"
+)
+
+const userAgentHeader = "User-Agent"
+
+// DefaultUserAgent is used in UserAgentTransport when the agent is not set.
+// It defaults to the project name + the current committed version.
+var DefaultUserAgent = "cloud-sdk-go/" + Version
+
+var defaultBackoff = time.Second * 1
+
+// CustomTransport is the cloud-sdk-go custom transport which handles all the
+// add-ons for http request and responses. It supports:
+// * Adding a custom UserAgent header to all outgoing requests.
+// * Writing a trail of the request / response flow to a device (verbose).
+// * Adding support for request retries on timeout with a backoff period.
+type CustomTransport struct {
```

Comment:
Is there an use case where this struct would be used outside of this package? If not I think it should be unexported

Reply:
I made it public to be able to use it in `ecl` and the `internal-cloud-sdk-go`